### DISCUSSION TITLE
How to get all included files used by batchcmds:compile(file.cpp) command to fill a batchcmds:add_depfiles(h1.hpp, h2.hpp....) ?

### DISCUSSION THREAD

**Discussion by PierreEVEN** [xmake-io/xmake#5946] [Q&A] [answered]
Hi,

I need some help to understand depfiles and batchcmds.

I'm building an UnrealEngine-like header tool that generates reflection headers.
Example : 
I create a class named `foo` inside a file `foo.hpp`, then my tool will parse this file, then generate a class in 2 generated files `foo.gen.cpp` and `foo.gen.hpp`

To achieve this, I'm using a custom xmake rule that will compile the generated `.gen.cpp` once updated :
```lua
rule("generated_cpp", function (rule)
    set_extensions(".hpp")
    before_buildcmd_file(function (target, batchcmds, source_header, opt)

        -- Guess generated source file path
        local generated_path = string.sub(os.projectdir().."/"..source_header, string.len(target:scriptdir()) + 2)
        -- replace .hpp extension with .gen.cpp
        local generated_source = target:autogendir().."/"..string.sub(generated_path, 1, string.len(generated_path) - 3).."gen.cpp"
        -- generated classes are always private
        generated_source = generated_source:gsub("public", "private", 1)
        
        -- check if the header tool generated reflection data for our input file
        if (os.exists(generated_source)) then
            local test_str = tostring(generated_source)
            local objectfile = target:objectfile(generated_source)
            --batchcmds:vrunv(TODO : make the header tool works for one header at a time)
            batchcmds:compile(generated_source, objectfile)

            batchcmds:show_progress(opt.progress, "${color.build.object}Compiling.reflection %s", generated_source)

            batchcmds:add_depfiles(source_header) -- The problem is here (explanations bellow)

            batchcmds:set_depmtime(os.mtime(objectfile))
            batchcmds:set_depcache(target:dependfile(objectfile))
        end
    end)
end)
```

My problem occurs when the input header file includes other headers : xmake will not track updates of subdependencies
example :
```cpp
// foo.hpp
#include "foo.gen.hpp"

class foo {}
```

```cpp
// bar.hpp
#include "foo.hpp" // if I update "foo.hpp", "bar.gen.cpp" will not be recompiled
#include "bar.gen.hpp"

class bar : public foo {}
```

Using the previous rule, xmake won't be able to know when `foo.hpp` is updated to recompile `bar.hpp` because `batchcmds:add_depfiles()` only includes `bar.hpp`

However, I've seen that xmake is able to parse includes somehow and build a complete recursive list of dependencies for each class, so I guess there is a way to solve my problem by adding the list of all dependencies recursively included by my source file in `batchcmds:add_depfiles()`.
But for now I have no clue how to do it.

Thanks in advance for your help !

**Comment by waruqi** [comment]
https://github.com/xmake-io/xmake/blob/46b99517934618116d6b280ffac819ba830af7db/xmake/rules/swig/xmake.lua#L119

https://github.com/xmake-io/xmake/blob/46b99517934618116d6b280ffac819ba830af7db/xmake/rules/swig/build_module_file.lua#L163-L187

https://github.com/xmake-io/xmake/blob/46b99517934618116d6b280ffac819ba830af7db/xmake/modules/private/action/build/object.lua#L91-L114

**Comment by PierreEVEN** [comment]
Thanks for the hint !
I'm not sure if I understood correctly, but if I provide a dependinfo object to `compile()`, it should fill the required dependency list ?
I've tried using `batchcmds:compile` but it doesn't do anything to the dependinfo object.
Also I don't know how to get the sourcekind in my case (opt is empty in a `before_buildcmd_file callback`) so I don't know how to retrieve a `compinst` like in your example. I assume `batchcmds:compile` do the same job, but I may be wrong ?

(I'm not familiar with lua at all, so I may have missed something obvious ?)

**Comment by PierreEVEN** [comment]
Update : I finally figured out how to use the compiler instance ! 🥳 

```cpp
import("core.tool.compiler")
local compinst = compiler.load("cxx", {target = target})
local compflags = compinst:compflags({target = target, sourcefile = generated_source, configs = opt.configs})
assert(compinst:compile(generated_source, objectfile, {dependinfo = dependinfo, compflags = compflags}))
```
So that worked, now I just need a way to find a way to make it works seamlessly with my batchcmd 👍

**Comment by PierreEVEN** [comment, marked as answer]
Yay found the solution 🥳
Thanks @waruqi !

```lua
rule("generated_cpp", function (rule)
    set_extensions(".hpp")
    before_buildcmd_file(function (target, batchcmds, source_header, opt)

        -- Guess generated source file path
        local generated_path = string.sub(os.projectdir().."/"..source_header, string.len(target:scriptdir()) + 2)
        -- replace .hpp extension with .gen.cpp
        local generated_source = target:autogendir().."/"..string.sub(generated_path, 1, string.len(generated_path) - 3).."gen.cpp"

        -- generated classes are always private
        generated_source = generated_source:gsub("public", "private", 1)
        
        if (os.exists(generated_source)) then
            local objectfile = target:objectfile(generated_source)
            --batchcmds:vrunv(TODO : make the header tool works for one header at a time)

            import("core.tool.compiler")
            import("core.project.depend")

            local compinst = compiler.load("cxx", {target = target})
            local compflags = compinst:compflags({target = target, sourcefile = generated_source, configs = opt.configs})


            local dependfile = target:dependfile(objectfile)
            -- Load existing dep infos (or create if not exists)
            local dependinfo = target:is_rebuilt() and {} or (depend.load(dependfile, {target = target}) or {})

            local depvalues = {compinst:program(), compflags}
            local lastmtime = os.isfile(objectfile) and os.mtime(objectfile) or 0

            -- Test if source file was modified
            if not depend.is_changed(dependinfo, {lastmtime = lastmtime, values = depvalues}) then
                return
            end

            batchcmds:show_progress(opt.progress, "${color.build.object}Compiling.Reflection %s", generated_source)
            -- Compile and fill the dependency list into dependinfo
            assert(compinst:compile(generated_source, objectfile, {dependinfo = dependinfo, compflags = compflags}))
            
            -- store build depvalues to detect depvalues changes
            dependinfo.values = depvalues
            depend.save(dependinfo, dependfile)
        end
    end)
end)
```

**Reply by waruqi** [reply]
batchcmd is imperative, and is usually used in project generators, such as xpack and `xmake project -k cmake/compile_commands/makefile/...`. It cannot maintain state and does not support parallelism.

If it is some simple rules, you can use batchcmd to simplify the configuration. If it is a more complex rule, it is recommended to use before_build_file instead of before_buildcmd_file.

It is more flexible and supports parallelism.

Of course, before_build_file/before_buildcmd_file can be configured at the same time, and `xmake build` will give priority to the former. `xmake project` only uses the latter.
